client.hpp

```diff
@@ -13,11 +18,40 @@ class Client {
   Client(int socket_desc);
 public:
   Client(const std::string& address, const std::string& port);
-  void write(const char* data, const size_t& length);
   Client& operator<<(const std::string& data);
   Client& operator>>(std::string& result);
-  std::vector<char> read();
-  std::string read_str();
+  std::string read_str() const;
+
+  template<typename T>
+  std::vector<T> read() const {
+    int len = 0;
+    ioctl(*m_socket_ptr, FIONREAD, &len);
+    if (len == 0) {
+      return std::vector<T>();
+    }
+    std::vector<T> result(len);
+    if (recv(*m_socket_ptr,(char *) result.data(), len, 0) == -1) {
+      throw std::runtime_error(strerror(errno));
+    }
+    return result;
+  }
+
+  template<typename T>
+  void write(const T& data, size_t length) {
+    if (send(*m_socket_ptr, data, length, 0) == -1) {
+      throw std::runtime_error(strerror(errno));
+    }
+  }
+
+  template<class T>
+  void write(const T& data) {
+    write(data.data(), data.size()); 
```

Comment:
Converting size to length may not always be correct. 
Consider this:
```
std::vector<uint32_t> vec{1u};
client << vec;
```
calls `client.write(&vec[0], 1); `, which writes one byte instead of one vector element. Send size is in bytes, so `data.size()` should be multiplied by size of one element of T(assuming T is a container).